src/main/java/org/mockito/plugins/DoNotMockEnforcer.java

```diff
@@ -4,20 +4,106 @@
  */
 package org.mockito.plugins;
 
+import org.mockito.DoNotMock;
+import org.mockito.NotExtensible;
+import org.mockito.mock.MockCreationSettings;
+
+import java.lang.annotation.Annotation;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
 /**
  * Enforcer that is applied to every type in the type hierarchy of the class-to-be-mocked.
  */
 public interface DoNotMockEnforcer {
 
     /**
-     * If this type is allowed to be mocked. Return an empty optional if the enforcer allows
+     * If this type is allowed to be mocked. Return {@code null} if the enforcer allows
      * this type to be mocked. Return a message if there is a reason this type can not be mocked.
-     *
+     * <p>
      * Note that Mockito performs traversal of the type hierarchy. Implementations of this class
      * should therefore not perform type traversal themselves but rely on Mockito.
      *
      * @param type The type to check
-     * @return Optional message if this type can not be mocked, or an empty optional if type can be mocked
+     * @return Optional message if this type can not be mocked, or {@code null} otherwise
+     *
+     * @see #checkTypeForDoNotMockViolation(MockCreationSettings)
+     */
+    default String checkTypeForDoNotMockViolation(Class<?> type) {
```

Comment:
I am not sure if I follow your logic why you moved the implementation from `DefaultDoNotMockEnforcer` into the interface? Since we want this interface to only expose 1 method to implement (`String checkTypeForDoNotMockViolation(Class<?>)`), I don't think we should implement it as default here.

E.g. my thinking was:

```java
public class DefaultDoNotMockEnforcer implements DoNotMockEnforcer {
    @Override
    public String checkTypeForDoNotMockViolation(Class<?> type) {
        // Implementation here
    }
}
```
```java
public interface DoNotMockEnforcer extends DoNotMockEnforcerWithType {
    String checkTypeForDoNotMockViolation(Class<?> type);

    default String checkTypeForDoNotMockViolation(MockCreationSettings<?> creationSettings)  {
        // Implementation with caching here
    }
}
```
```java
public interface DoNotMockEnforcerWithType {
    checkTypeForDoNotMockViolation(MockCreationSettings<?> creationSettings);
}
```

I understand that our plugin system currently expects `DoNotMockEnforcer`, but I think we can support both here. E.g. we add https://github.com/mockito/mockito/blob/3db5df66858ba94aa5776bae97becac62ee45407/src/main/java/org/mockito/internal/configuration/plugins/DefaultMockitoPlugins.java#L65 to reference `DoNotMockEnforcerWithType`. Then when we choose which mockenforcer to load, we first check for `DoNotMockEnforcer` and `DoNotMockEnforcerWithType` and we check if one exists. Otherwise we choose the default implementation. For that logic, we can use the existing `alternatePluginType` in https://github.com/mockito/mockito/blob/3db5df66858ba94aa5776bae97becac62ee45407/src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java#L63 In other words, update https://github.com/mockito/mockito/blob/3db5df66858ba94aa5776bae97becac62ee45407/src/main/java/org/mockito/internal/configuration/plugins/PluginRegistry.java#L50 to

```java
new PluginLoader(pluginSwitch).loadPlugin(DoNotMockEnforcerWithType.class, DoNotMockEnforcer.class);
```

Reply:
Do we have other instances of different plugins that might take precedence over another?

My thinking was that having two plugins would either result in code duplication, or create some sort of dependency between plugins. What you suggest would be some sort of order of precedence, but that's still potentially confusing (and it would still involve one of the two options mentioned before). I guess we'd have to add validations against users customizing both plugin types at the same time, among other potentially problematic scenarios?

I guess there's no easy solution for the problem, but at least keeping everything inside `DoNotMockEnforcer` allows one to keep the existing implementations working and avoid the hassle of dealing with cross-plugin dependencies, duplication and delving into validations that would be expected from a sort of order of precedence kind of thing. And it's all in one place, which makes it potentially easier to reason about and reuse in their own implementations (e.g. they wouldn't need to inherit/reference `DefaultDoNotMockEnforcer`, which is internal, in order to maintain the rest of their custom plugin implementation compatible with the annotation without having to copy-paste the code in their own plugin implementation).

I think having everything in the same interface gives us the best of the customizability and reuse while allowing users to utilize existing code without making references to internal classes.

Reply:
@TimvdLippe I gave it some more thought, but I don't think having this order of precedence is a good idea.

The idea was to allow the users to customize the checks more deeply than simply checking each type individually, by allowing them access to the traversal order. We cannot accomplish that with a system that has both `DoNotMockEnforcerWithType` and `DoNotMockEnforcer` in a order-of-precedence relationship, unless we severely duplicate code, which won't be very nice, IMHO.

Right now, we have:

1. Traversal done in `MockitoCore`
2. Each check passes through `DoNotMockEnforcer`
3. Users must reference `DefaultDoNotMockEnforcer` in order to gain access to the default behavior, in case they want to simply extend Mockito's behavior and benefit from future developments. Otherwise, they're not able to use the default behavior and must copy-paste the code from `DefaultDoNotMockEnforcer` onto theirs.

With the proposed precedence of `DoNotMockEnforcerWithType` over `DoNotMockEnforcer` default implementation of `DoNotMockEnforcerWithType`, as that would inevitably override even user's own custom `DoNotMockEnforcer`. If we have the precedence over to `DoNotMockEnforcer`, then we'll never load `DoNotMockEnforcerWithType` either, because we have a default `DoNotMockEnforcer`. 

So, if having `DoNotMockEnforcerWithType` is desired, they must not take precedence over one another. This would lead, though, to either making the default `DoNotMockEnforcerWithType` depend on `DoNotMockEnforcer`, or duplicating the code of the latter onto the former.  This would still make extending the default behavior necessitate referencing an internal class, also.

The only way to avoid the dependency and/or duplication is to keep the traversal internal (e.g. `MockitoCore`) and have `MockitoCore` rely on both `DoNotMockEnforcerWithType` and `DoNotMockEnforcer`. Make `DoNotMockEnforcerWithType` have no default implementation and, if and only if one exists, and then had over the traversal to it.

The obvious downsides are:

- Two extension points for essentially a single function
- Still impossible to extend the current behavior, at least without relying on referencing internal classes (for `DoNotMockEnforcer`). For `DoNotMockEnforcerWithType`, it'd be necessary to copy the traversal implementation from `MockitoCore` onto one's own custom one. 

I think moving the implementation onto `DoNotMockEnforcer` makes it both simpler and easier to extend without referencing any internal classes, not to mention all behavior related to this check stays in a single place, and it's 100% backwards compatible.

Reply:
I don't love an interface with two default methods, but I think it's a lot better than the alternatives.

I'd try to make the caching cleaner, though. Right now it works but it's kinda ugly.

Reply:
Thanks for the detailed write-up. I think from my side, the following would be ideal:

1. We retain the existing do-not-mock mechanism for users. We can't change the behavior, as that would be a breaking change
2. We retain an interface named `DoNotMockEnforcer` that has one method to be implemented, as that is the current public API we expose
3. We want to expose an override to allow for more specific type checking. When users implement the overrideable interface, it should take precedence over the default implementation
4. We want to expose the existing implementation for efficient reuse, but not necessarily as part of the interface definition

Given that we currently ship an implementation of `DoNotMockEnforcer` and if `DoNotMockEnforcer` would extend `DoNotMockEnforcerWithType`, the precedence to me is that we first attempt to see if we have a donotmock extension that is named `DoNotMockEnforcerWithType`. If that exists, that means the user ships their own implementation and thus we need to use that. If they don't, then it would check for `DoNotMockEnforcer`. If the user ships their own implementation, again it would use that. If either of these fail, we load `DefaultDoNotMockEnforcer`.

We currently have none of these precedence implementations, as we removed these in previous major version bumps. However, we have used this mechanism before to ship backwards compatible interface changes to our plugin system.

I understand that we are currently deep into implementation detail, but would it maybe be good to start writing conformance tests for the above mentioned scenarios so that we know what we are working towards? That would help us to eventually validate that whatever solution we decide to go for implements it how we want it. This allows us to figure out if every requirement we currently have defined makes sense. I recommend creating a new subproject for these things so that we can even use the plugin mechanism directly, similar to what we have done with https://github.com/mockito/mockito/blob/main/subprojects/junitJupiterInlineMockMakerExtensionTest/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker

In this case, we have one subproject that creates a `subprojects/doNotMockMakerTest/src/test/resources/mockito-extensions/org.mockito.DoNotMockEnforcer` and another subproject that creates `subprojects/doNotMockMakerWithTypeTest/src/test/resources/mockito-extensions/org.mockito.DoNotMockEnforcerWithType`. Note that the first one implements a plugin extension for the existing public API, whereas the other one implements only the new one. That way, we know for sure the user decides which one is which.

With regards to your code reuse point: I agree that I do think it is beneficial for our users to reuse parts of that. However, I wouldn't ship that as part of default methods, but instead ship that part of a utility method. For that, I think we can ship a `static` method on the interface that implements the utility.

Reply:
Yes, this sounds good to me. I think I misunderstood you before. I'm gonna restructure it, review the changes and build the conformance test.

Because the conformance test case was clear to me, I ended up going straight into the implementation details, though I was mostly worried about what it would look like in terms of backwards compatibility and extensibility, and your solution does both, so I'm happy with implementing it. 

I can have them as static methods in the interface, but is that too different than having a default implementation for the method a user can just call with `DoNotMockEnforcer.super. checkTypeForDoNotMockViolation(type)`? The static method will work, but I guess default methods feel more "natural"? At this point, I'm just asking out of curiosity, in case there are technical reasons to prefer one over another.

Reply:
Oh, I realized now why I did what I did. It's not possible to provide a default implementation of `DoNotMockEnforcerWithType` without referencing `DoNotMockEnforcer`. 

While I could make them both refer to a common default static implementation, the problem becomes that the default `DoNotMockEnforcerWithType` implementation must be able to use the custom `DoNotMockEnforcer` implementations as well.

The closest I could make is to have:

```java
public interface DoNotMockEnforcerWithType {

    String checkTypeForDoNotMockViolation(Class<?> type);

    String checkTypeForDoNotMockViolation(MockCreationSettings<?> creationSettings);
}
```

```java
public interface DoNotMockEnforcer extends DoNotMockEnforcerWithType {
    @Override
    default String checkTypeForDoNotMockViolation(MockCreationSettings<?> creationSettings) {
        // implementation and maybe caching or something, perhaps on a static utility method somewhere so it can be reused
    }
}
```

I think it just ends up converging back to what I proposed, because the rest either fails to accomplish the reuse aspect, or makes the situation worse by forcing this sort of unnatural relationship between classes.